pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,180 @@
+type reconciler struct {
+	client    client.Client
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.client.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// List all the Namespaces which matches the Namespace LabelSelector.
+	namespaceList := corev1.NamespaceList{}
+	if ingressController.Spec.NamespaceSelector != nil {
+		namespaceLabelSelector := ingressController.Spec.NamespaceSelector.MatchLabels
+		if err := r.client.List(ctx, &namespaceList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(namespaceLabelSelector),
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %v", request, err)
+		}
+	} else {
+		// If Namespace LabelSelector is not provided then add Namespace with empty name to indicate all Namespaces.
+		namespaceList.Items = append(namespaceList.Items, corev1.Namespace{})
+	}
+
+	// Get the Route LabelSelector from the Ingress Controller.
+	var routeLabelSelector map[string]string
+	if ingressController.Spec.RouteSelector != nil {
+		routeLabelSelector = ingressController.Spec.RouteSelector.MatchLabels
+	}
+
+	// Store the previous value of the metric.
+	prevMetricValue := GetRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Initialize the RouteMetricsControllerRoutesPerShard metric for Ingress Controller.
+	InitializeRouteMetricsControllerRoutesPerShardMetric(request.Name)
+
+	// Iterate through all the Namespaces.
+	for _, namespace := range namespaceList.Items {
+		// List all the Routes which matches the Route LabelSelector and the Namespace.
+		routeList := routev1.RouteList{}
+		if err := r.client.List(ctx, &routeList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(routeLabelSelector),
+			Namespace:     namespace.Name,
+		}); err != nil {
+			// Set the value of the metric to its previous value and re-queue the request.
+			SetRouteMetricsControllerRoutesPerShardMetric(request.Name, prevMetricValue)
+			return reconcile.Result{}, fmt.Errorf("failed to list Routes for the Shard %q: %v", request, err)
```

Comment:
There could be routeLabelSelectors without any matching Routes, so this is not really an error.

Reply:
If there are no routes found then a RouteList would be returned with an empty `Items` slice. It'll not return any error for that. I have tested this scenario locally.